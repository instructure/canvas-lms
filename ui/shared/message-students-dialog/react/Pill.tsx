/*
 * Copyright (C) 2022 - present Instructure, Inc.
 *
 * This file is part of Canvas.
 *
 * Canvas is free software: you can redistribute it and/or modify it under
 * the terms of the GNU Affero General Public License as published by the Free
 * Software Foundation, version 3 of the License.
 *
 * Canvas is distributed in the hope that it will be useful, but WITHOUT ANY
 * WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR
 * A PARTICULAR PURPOSE. See the GNU Affero General Public License for more
 * details.
 *
 * You should have received a copy of the GNU Affero General Public License along
 * with this program. If not, see <http://www.gnu.org/licenses/>.
 */

import React from 'react'
import {Tag} from '@instructure/ui-tag'
import {Text} from '@instructure/ui-text'
import {Tooltip} from '@instructure/ui-tooltip'
import {Flex} from '@instructure/ui-flex'
import {IconAddSolid, IconXSolid} from '@instructure/ui-icons'
import {ApplyTheme} from '@instructure/ui-themeable'
import {useScope as useI18nScope} from '@canvas/i18n'
import {TruncateText} from '@instructure/ui-truncate-text'

const themeOverride = {
  [Tag.theme]: {
    defaultBackground: 'white',
  },
}

const I18n = useI18nScope('pill')
const ellipsis = () => I18n.t('â€¦')
const truncate = text => (text.length > 14 ? text.slice(0, 13) + ellipsis() : text)

function renderText(text, truncatedText, textColor) {
  const isTruncated = text.length > truncatedText.length
  if (isTruncated) {
    return (
      <Tooltip renderTip={text}>
        <Text as="div" color={textColor}>
          <TruncateText>{truncatedText}</TruncateText>
        </Text>
      </Tooltip>
    )
  } else {
    return (
      <Text as="div" color={textColor}>
        <TruncateText>{text}</TruncateText>
      </Text>
    )
  }
}

function renderIcon(selected) {
  if (selected) {
    return <IconXSolid data-testid="item-selected" />
  } else {
    return <IconAddSolid data-testid="item-unselected" color="brand" />
  }
}

const Pill = ({studentId, observerId = null, text, onClick, selected = false}) => {
  const textColor = selected ? 'primary' : 'secondary'
  const truncatedText = truncate(text)

  const contents = (
    <Flex as="div" margin="0 xx-small 0 0" justifyItems="space-between">
      <Flex.Item size="0.75rem" shouldGrow={true} margin="0 xx-small 0 0" overflowX="hidden">
        {renderText(text, truncatedText, textColor)}
      </Flex.Item>
      <Flex.Item>{renderIcon(selected)}</Flex.Item>
    </Flex>
  )

  return (
    <ApplyTheme theme={themeOverride}>
      <Tag text={contents} onClick={() => onClick(studentId, observerId)} />
    </ApplyTheme>
  )
}

export default Pill
