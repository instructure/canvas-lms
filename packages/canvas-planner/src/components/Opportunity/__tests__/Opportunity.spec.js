/*
 * Copyright (C) 2017 - present Instructure, Inc.
 *
 * This file is part of Canvas.
 *
 * Canvas is free software: you can redistribute it and/or modify it under
 * the terms of the GNU Affero General Public License as published by the Free
 * Software Foundation, version 3 of the License.
 *
 * Canvas is distributed in the hope that they will be useful, but WITHOUT ANY
 * WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR
 * A PARTICULAR PURPOSE. See the GNU Affero General Public License for more
 * details.
 *
 * You should have received a copy of the GNU Affero General Public License along
 * with this program. If not, see <http://www.gnu.org/licenses/>.
 */
import React from 'react';
import { shallow, mount } from 'enzyme';
import { Opportunity } from '../index';
import Pill from '@instructure/ui-core/lib/components/Pill';

function defaultProps (options) {
  return {
    id: "1",
    dueAt: "2017-03-09T20:40:35Z",
    courseName: "course about stuff",
    opportunityTitle: "this is a description about the opportunity",
    points: 20,
    showPill: true,
    url: "http://www.non_default_url.com",
    timeZone: 'America/Denver',
    dismiss: () => {},
    registerAnimatable: () => {},
    deregisterAnimatable: () => {},
    animatableIndex: 1,
  };
}

it('renders the base component correctly', () => {
  const wrapper = shallow(
    <Opportunity {...defaultProps()} />
  );
  expect(wrapper).toMatchSnapshot();
});

it('calls the onClick prop when dismissed is clicked', () => {
  let tempProps = defaultProps();
  tempProps.dismiss = jest.fn();
  const wrapper = shallow(
    <Opportunity {...tempProps}/>
  );
  wrapper.find('Button').simulate('click');
  expect(tempProps.dismiss).toHaveBeenCalled();
});

it('renders the base component correctly without points', () => {
  let tempProps = defaultProps();
  tempProps.points = null;
  const wrapper = shallow(
    <Opportunity {...tempProps} />
  );
  expect(wrapper).toMatchSnapshot();
});

it('renders a Pill if passed showPill: true', () => {
  const props = defaultProps();
  const wrapper = shallow(<Opportunity {...props} />);
  expect(wrapper.find(Pill).length).toEqual(1);
});

it('does not render a Pill if passed showPill: false', () => {
  const props = defaultProps();
  props.showPill = false;
  const wrapper = shallow(<Opportunity {...props} />);
  expect(wrapper.find(Pill).length).toEqual(0);
});

it('registers itself as animatable', () => {
  const fakeRegister = jest.fn();
  const fakeDeregister = jest.fn();
  const wrapper = mount(
    <Opportunity
      {...defaultProps()}
      id='1'
      registerAnimatable={fakeRegister}
      deregisterAnimatable={fakeDeregister}
      animatableIndex={42}
    />
  );
  const instance = wrapper.instance();
  expect(fakeRegister).toHaveBeenCalledWith('opportunity', instance, 42, ['1']);

  wrapper.setProps({id: '2', animatableIndex: 43});
  expect(fakeDeregister).toHaveBeenCalledWith('opportunity', instance, ['1']);
  expect(fakeRegister).toHaveBeenCalledWith('opportunity', instance, 43, ['2']);

  wrapper.unmount();
  expect(fakeDeregister).toHaveBeenCalledWith('opportunity', instance, ['2']);
});
