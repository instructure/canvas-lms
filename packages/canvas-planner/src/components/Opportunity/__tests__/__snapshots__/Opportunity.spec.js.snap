// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders the base component correctly 1`] = `
<div>
  <div>
    <div>
      course about stuff
    </div>
    <div>
      <Link
        as="button"
        ellipsis={false}
        href="http://www.non_default_url.com"
        iconPlacement="start"
        linkRef={[Function]}
        variant="default"
      >
        this is a description about the opportunity
      </Link>
    </div>
  </div>
  <div>
    <div>
      <Pill
        text="Missing"
        variant="danger"
      />
      <div>
        <span>
          Due:
        </span>
         
        March 9, 2017 at 1:40 PM
      </div>
    </div>
    <div>
      <ScreenReaderContent
        as="span"
      >
        20 points
      </ScreenReaderContent>
      <PresentationContent
        as="span"
      >
        <span>
          20
        </span>
        points
      </PresentationContent>
    </div>
  </div>
  <div>
    <Button
      as="button"
      buttonRef={[Function]}
      fluidWidth={false}
      margin="0"
      onClick={[Function]}
      size="small"
      type="button"
      variant="icon"
    >
      <IconX
        title="Dismiss this is a description about the opportunity"
      />
    </Button>
  </div>
</div>
`;

exports[`renders the base component correctly without points 1`] = `
<div>
  <div>
    <div>
      course about stuff
    </div>
    <div>
      <Link
        as="button"
        ellipsis={false}
        href="http://www.non_default_url.com"
        iconPlacement="start"
        linkRef={[Function]}
        variant="default"
      >
        this is a description about the opportunity
      </Link>
    </div>
  </div>
  <div>
    <div>
      <Pill
        text="Missing"
        variant="danger"
      />
      <div>
        <span>
          Due:
        </span>
         
        March 9, 2017 at 1:40 PM
      </div>
    </div>
    <div>
      <ScreenReaderContent
        as="span"
      >
        null points
      </ScreenReaderContent>
      <PresentationContent
        as="span"
      >
        <span />
        points
      </PresentationContent>
    </div>
  </div>
  <div>
    <Button
      as="button"
      buttonRef={[Function]}
      fluidWidth={false}
      margin="0"
      onClick={[Function]}
      size="small"
      type="button"
      variant="icon"
    >
      <IconX
        title="Dismiss this is a description about the opportunity"
      />
    </Button>
  </div>
</div>
`;
